File: 0113-path-sum-ii/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
       List<List<Integer>> result = new ArrayList<>();
       List<Integer> paths = new ArrayList<>();

       // Edge case 
       if(root == null) {
         return result;
       }  

       dfs(root,targetSum,paths,result);

       return result;
        
    }

    private void dfs(
        TreeNode node,
        int targetSum,
        List<Integer> paths,
        List<List<Integer>> result
    ) {

        //add the current node val to the paths
        paths.add(node.val);
        targetSum = targetSum - node.val;

        // Base case (if the current node is the leaf)

         if(node.left == null && node.right == null) {
             if(targetSum == 0) {
                result.add(new ArrayList(paths));
             }
         }

        //Recursive case 

         if(node.left != null) {
            dfs(node.left,targetSum,paths,result);
         }

         if(node.right != null) {
            dfs(node.right,targetSum,paths,result);
         }

        paths.remove(paths.size()-1); 

    }
}

File: 0113-path-sum-ii/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if root is None:
            return []
        result = []

        def dfs(node, target, slate):
            slate.append(node.val)
            target = target - node.val

            # BaseCase
            if node.left is None and node.right is None:
                # Append the copy
                if target == 0:
                  result.append(slate[:])

            # Recursive Case
            if node.left is not None:
                dfs(node.left, target, slate)

            if node.right is not None:
                dfs(node.right, target, slate)

            # you have to pop the slate to remove the last state
            slate.pop()

        dfs(root, targetSum, [])
        return result

